service/src/main/java/bio/terra/catalog/service/DatasetService.java

```diff
@@ -55,49 +56,76 @@ public IllegalMetadataException(Throwable cause) {
+      } catch (JsonProcessingException e) {
+        // This shouldn't occur, as the data stored in postgres must be valid JSON, because it's
+        // stored as JSONB.
+        throw new IllegalMetadataException(e);
+      }
     }
   }
 
-  private List<ObjectNode> convertSourceObjectsToDatasetResponses(
+  private List<DatasetWithAccessLevel> convertSourceObjectsToDatasetsWithAccessLevel(
       Map<String, DatasetAccessLevel> roleMap, StorageSystem storageSystem) {
     List<Dataset> datasets = datasetDao.find(storageSystem, roleMap.keySet());
     return datasets.stream()
-        .map(dataset -> sourceAndDatasetToObjectNode(roleMap, dataset))
+        .map(dataset -> new DatasetWithAccessLevel(dataset, roleMap.get(dataset.storageSourceId())))
         .toList();
   }
 
-  private ObjectNode sourceAndDatasetToObjectNode(
-      Map<String, DatasetAccessLevel> roleMap, Dataset dataset) {
-    ObjectNode node = toJsonNode(dataset.metadata());
-    node.set(
-        "accessLevel", TextNode.valueOf(String.valueOf(roleMap.get(dataset.storageSourceId()))));
-    node.set("id", TextNode.valueOf(dataset.id().toValue()));
-    return node;
-  }
-
-  private List<ObjectNode> collectDatarepoDatasets(AuthenticatedUserRequest user) {
+  private List<DatasetWithAccessLevel> collectDatarepoDatasets(AuthenticatedUserRequest user) {
     // For this storage system, get the collection of visible datasets and the user's roles for
     // each dataset.
     var roleMap = datarepoService.getSnapshotIdsAndRoles(user);
-    return convertSourceObjectsToDatasetResponses(roleMap, StorageSystem.TERRA_DATA_REPO);
+    return convertSourceObjectsToDatasetsWithAccessLevel(roleMap, StorageSystem.TERRA_DATA_REPO);
   }
 
-  private List<ObjectNode> collectWorkspaceDatasets(AuthenticatedUserRequest user) {
+  private List<DatasetWithAccessLevel> collectWorkspaceDatasets(AuthenticatedUserRequest user) {
     var roleMap = rawlsService.getWorkspaceIdsAndRoles(user);
-    return convertSourceObjectsToDatasetResponses(roleMap, StorageSystem.TERRA_WORKSPACE);
+    return convertSourceObjectsToDatasetsWithAccessLevel(roleMap, StorageSystem.TERRA_WORKSPACE);
   }
 
   public DatasetsListResponse listDatasets(AuthenticatedUserRequest user) {
+    List<DatasetWithAccessLevel> datasets = new ArrayList<>();
+    datasets.addAll(collectWorkspaceDatasets(user));
+    datasets.addAll(collectDatarepoDatasets(user));
+    if (samService.hasGlobalAction(user, SamAction.READ_ANY_METADATA)) {
+      datasets.addAll(
+          datasetDao.listAllDatasets().stream()
+              .map(dataset -> new DatasetWithAccessLevel(dataset, DatasetAccessLevel.READER))
+              .filter(
+                  datasetWithAccessLevel ->
+                      !datasets.stream()
+                          .map(datasetInList -> datasetInList.getDataset().id())
+                          .toList()
+                          .contains(datasetWithAccessLevel.getDataset().id()))
+              .toList());
```

Comment:
I think `x.addAll(y.stream().toList())` is less memory efficient than `y.stream().forEach(x::add)`